{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "f1712c0d",
   "metadata": {
    "code_folding": [
     20
    ]
   },
   "outputs": [],
   "source": [
    "## Export to html\n",
    "#  jupyter nbconvert course_network.ipynb --no-input --to html\n",
    "# \n",
    "\n",
    "import yaml, re, os, nbconvert, nbformat\n",
    "import pandas as pd\n",
    "import graphviz\n",
    "\n",
    "main_dir=re.sub('code', '', os.getcwd())\n",
    "data_dir=main_dir+'data/'\n",
    "results_dir=main_dir+'results/'\n",
    "\n",
    "with open(data_dir+'course_network.yaml') as file:\n",
    "    course_in=yaml.safe_load(file)\n",
    "    \n",
    "# Using `pygraphviz` for exporting. To export:\n",
    "# 1. jupyter nbconvert course_network.ipynb --no-input --to html\n",
    "# ```\n",
    "# <IMG SRC=\"/path/to/course_graph.svg\" USEMAP=\"#my_graph\" />\n",
    "# ```\n",
    "\n",
    "import pygraphviz as pgv\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "82e22301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Title: G Pages: 1 -->\n",
       "<svg width=\"1200pt\" height=\"476pt\"\n",
       " viewBox=\"0.00 0.00 1200.20 476.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 472)\">\n",
       "<title>G</title>\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-472 1196.2,-472 1196.2,4 -4,4\"/>\n",
       "<!-- Ph.D. Advisory Program -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>Ph.D. Advisory Program</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"61.2\" cy=\"-450\" rx=\"61.4\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"61.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">Ph.D. Advisory Program</text>\n",
       "</g>\n",
       "<!-- Advanced Python -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>Advanced Python</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"272.2\" cy=\"-342\" rx=\"46.53\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"272.2\" y=\"-339.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">Advanced Python</text>\n",
       "</g>\n",
       "<!-- Introduction to Python -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>Introduction to Python</title>\n",
       "<g id=\"a_node3\"><a xlink:href=\"https://uppsala.instructure.com/courses/47059\" xlink:title=\"Introduction to Python\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"197.2\" cy=\"-450\" rx=\"56.66\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"197.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\">Introduction to Python</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Advanced Python -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Advanced Python</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M209.25,-431.97C221.51,-414.64 240.65,-387.59 254.61,-367.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"260.43,-359.64 258.32,-370.4 257.54,-363.72 254.65,-367.8 254.65,-367.8 254.65,-367.8 257.54,-363.72 250.98,-365.2 260.43,-359.64 260.43,-359.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>Intro to biostatistics and ML</title>\n",
       "<g id=\"a_node6\"><a xlink:href=\"https://uppsala.instructure.com/courses/51998\" xlink:title=\"Intro to biostatistics and ML\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"405.2\" cy=\"-342\" rx=\"68.91\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"405.2\" y=\"-339.8\" font-family=\"Times,serif\" font-size=\"9.00\">Intro to biostatistics and ML</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Intro to biostatistics and ML -->\n",
       "<!-- Neural Nets and DL -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>Neural Nets and DL</title>\n",
       "<g id=\"a_node12\"><a xlink:href=\"https://uppsala.instructure.com/courses/47910\" xlink:title=\"Neural Nets and DL\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"324.2\" cy=\"-234\" rx=\"51.27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"324.2\" y=\"-231.8\" font-family=\"Times,serif\" font-size=\"9.00\">Neural Nets and DL</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Neural Nets and DL -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Neural Nets and DL</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M189.38,-432.01C179.02,-406.77 164.12,-358.61 183.2,-324 203.23,-287.66 245.04,-264.11 277.96,-250.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"287.42,-246.66 279.79,-254.54 282.77,-248.51 278.13,-250.36 278.13,-250.36 278.13,-250.36 282.77,-248.51 276.46,-246.18 287.42,-246.66 287.42,-246.66\"/>\n",
       "</g>\n",
       "<!-- Omics Integration -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>Omics Integration</title>\n",
       "<g id=\"a_node13\"><a xlink:href=\"https://uppsala.instructure.com/courses/52162\" xlink:title=\"Omics Integration\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"485.2\" cy=\"-234\" rx=\"47.19\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"485.2\" y=\"-231.8\" font-family=\"Times,serif\" font-size=\"9.00\">Omics Integration</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M194.63,-431.79C191.71,-405.38 190.29,-354.7 216.2,-324 244.55,-290.41 364.41,-260.35 434.2,-245.25\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"444.12,-243.14 435.28,-249.63 439.23,-244.18 434.34,-245.22 434.34,-245.22 434.34,-245.22 439.23,-244.18 433.4,-240.82 444.12,-243.14 444.12,-243.14\"/>\n",
       "</g>\n",
       "<!-- single cell RNAseq -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>single cell RNAseq</title>\n",
       "<g id=\"a_node16\"><a xlink:href=\"https://uppsala.instructure.com/courses/52011\" xlink:title=\"single cell RNAseq\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"670.2\" cy=\"-126\" rx=\"49.8\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"670.2\" y=\"-123.8\" font-family=\"Times,serif\" font-size=\"9.00\">single cell RNAseq</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M187.56,-431.98C175.02,-407.5 156.27,-361.22 170.2,-324 192.5,-264.4 209.46,-248.45 264.2,-216 374.58,-150.55 527.52,-133.09 610.63,-128.52\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"620.66,-128.01 610.9,-133.01 615.66,-128.27 610.67,-128.52 610.67,-128.52 610.67,-128.52 615.66,-128.27 610.44,-124.02 620.66,-128.01 620.66,-128.01\"/>\n",
       "</g>\n",
       "<!-- R Programming -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>R Programming</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"599.2\" cy=\"-450\" rx=\"42.95\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"599.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">R Programming</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Advanced Python -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Advanced Python</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M566.35,-438.35C507.35,-419.23 383.39,-379.04 316.78,-357.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"306.99,-354.28 317.9,-353.08 311.75,-355.82 316.51,-357.36 316.51,-357.36 316.51,-357.36 311.75,-355.82 315.12,-361.64 306.99,-354.28 306.99,-354.28\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Intro to biostatistics and ML -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Intro to biostatistics and ML</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M574.23,-435.36C540.79,-417.09 481.17,-384.51 442.42,-363.34\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"433.53,-358.48 444.46,-359.32 437.92,-360.88 442.3,-363.27 442.3,-363.27 442.3,-363.27 437.92,-360.88 440.15,-367.22 433.53,-358.48 433.53,-358.48\"/>\n",
       "</g>\n",
       "<!-- Data visualization with R -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>Data visualization with R</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"554.2\" cy=\"-342\" rx=\"62.21\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"554.2\" y=\"-339.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">Data visualization with R</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Data visualization with R -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Data visualization with R</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M591.97,-431.97C584.79,-415.05 573.67,-388.86 565.35,-369.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"561.41,-359.99 569.46,-367.43 563.37,-364.59 565.32,-369.19 565.32,-369.19 565.32,-369.19 563.37,-364.59 561.18,-370.95 561.41,-359.99 561.41,-359.99\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M609.56,-432.15C623.52,-407.09 644.51,-359.17 625.2,-324 605.07,-287.35 562.75,-263.65 529.88,-249.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"520.45,-246.24 531.41,-245.76 525.1,-248.09 529.74,-249.94 529.74,-249.94 529.74,-249.94 525.1,-248.09 528.08,-254.12 520.45,-246.24 520.45,-246.24\"/>\n",
       "</g>\n",
       "<!-- RaukR Summer course -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>RaukR Summer course</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"756.2\" cy=\"-234\" rx=\"57.97\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"756.2\" y=\"-231.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">RaukR Summer course</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M619.52,-434.15C640.75,-417.82 674.01,-389.93 696.2,-360 719.08,-329.14 736.75,-288.17 746.82,-261.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"750.3,-252.31 751.04,-263.25 748.56,-256.99 746.82,-261.68 746.82,-261.68 746.82,-261.68 748.56,-256.99 742.6,-260.12 750.3,-252.31 750.3,-252.31\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M609.63,-432.32C619.96,-414.95 635.34,-386.61 643.2,-360 664.52,-287.82 669.13,-198.4 670.04,-154.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"670.2,-144.06 674.55,-154.13 670.13,-149.06 670.05,-154.06 670.05,-154.06 670.05,-154.06 670.13,-149.06 665.55,-153.99 670.2,-144.06 670.2,-144.06\"/>\n",
       "</g>\n",
       "<!-- Bioinformatics for PIs -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>Bioinformatics for PIs</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1136.2\" cy=\"-450\" rx=\"56.01\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1136.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">Bioinformatics for PIs</text>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;Neural Nets and DL -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;Neural Nets and DL</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M392.19,-323.97C378.88,-306.56 358.08,-279.34 342.99,-259.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"336.91,-251.64 346.56,-256.85 339.95,-255.61 342.99,-259.58 342.99,-259.58 342.99,-259.58 339.95,-255.61 339.41,-262.32 336.91,-251.64 336.91,-251.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M418.05,-323.97C431.13,-306.64 451.55,-279.59 466.44,-259.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"472.64,-251.64 470.21,-262.33 469.63,-255.63 466.62,-259.62 466.62,-259.62 466.62,-259.62 469.63,-255.63 463.03,-256.91 472.64,-251.64 472.64,-251.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M448.59,-327.9C513.97,-308.15 637.42,-270.87 706.39,-250.04\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"716.16,-247.09 707.88,-254.29 711.37,-248.54 706.58,-249.98 706.58,-249.98 706.58,-249.98 711.37,-248.54 705.28,-245.68 716.16,-247.09 716.16,-247.09\"/>\n",
       "</g>\n",
       "<!-- Epigenomics data analysis -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>Epigenomics data analysis</title>\n",
       "<g id=\"a_node8\"><a xlink:href=\"https://uppsala.instructure.com/courses/51975\" xlink:title=\"Epigenomics data analysis\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"896.2\" cy=\"-234\" rx=\"64.83\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"896.2\" y=\"-231.8\" font-family=\"Times,serif\" font-size=\"9.00\">Epigenomics data analysis</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>Intro to NGS</title>\n",
       "<g id=\"a_node9\"><a xlink:href=\"https://uppsala.instructure.com/courses/47037\" xlink:title=\"Intro to NGS\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"852.2\" cy=\"-342\" rx=\"36.25\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"852.2\" y=\"-339.8\" font-family=\"Times,serif\" font-size=\"9.00\">Intro to NGS</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;Epigenomics data analysis -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;Epigenomics data analysis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M859.27,-323.97C866.29,-307.05 877.16,-280.86 885.29,-261.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"889.15,-251.99 889.47,-262.95 887.23,-256.61 885.31,-261.22 885.31,-261.22 885.31,-261.22 887.23,-256.61 881.16,-259.5 889.15,-251.99 889.15,-251.99\"/>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M821.54,-332.14C757.16,-313.55 607.57,-270.34 531.46,-248.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"521.67,-245.53 532.52,-243.98 526.47,-246.92 531.27,-248.31 531.27,-248.31 531.27,-248.31 526.47,-246.92 530.02,-252.63 521.67,-245.53 521.67,-245.53\"/>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M838.03,-325.36C822.19,-307.87 796.35,-279.34 777.95,-259.02\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"771.05,-251.4 781.1,-255.79 774.41,-255.11 777.77,-258.81 777.77,-258.81 777.77,-258.81 774.41,-255.11 774.43,-261.84 771.05,-251.4 771.05,-251.4\"/>\n",
       "</g>\n",
       "<!-- RNAseq data analysis -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>RNAseq data analysis</title>\n",
       "<g id=\"a_node15\"><a xlink:href=\"https://uppsala.instructure.com/courses/52164\" xlink:title=\"RNAseq data analysis\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1034.2\" cy=\"-234\" rx=\"55.35\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1034.2\" y=\"-231.8\" font-family=\"Times,serif\" font-size=\"9.00\">RNAseq data analysis</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;RNAseq data analysis -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;RNAseq data analysis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M874.88,-327.79C906.15,-309.58 962.79,-276.59 999.43,-255.25\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1008.29,-250.09 1001.91,-259.01 1003.97,-252.61 999.65,-255.12 999.65,-255.12 999.65,-255.12 1003.97,-252.61 997.38,-251.23 1008.29,-250.09 1008.29,-250.09\"/>\n",
       "</g>\n",
       "<!-- Intro to Data Management -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>Intro to Data Management</title>\n",
       "<g id=\"a_node10\"><a xlink:href=\"https://uppsala.instructure.com/courses/52280\" xlink:title=\"Intro to Data Management\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"823.2\" cy=\"-450\" rx=\"64.83\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"823.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\">Intro to Data Management</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to Data Management&#45;&gt;Intro to NGS -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>Intro to Data Management&#45;&gt;Intro to NGS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M827.86,-431.97C832.45,-415.2 839.52,-389.34 844.86,-369.81\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"847.55,-359.99 849.25,-370.82 846.23,-364.81 844.91,-369.63 844.91,-369.63 844.91,-369.63 846.23,-364.81 840.57,-368.45 847.55,-359.99 847.55,-359.99\"/>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>Tools for Reproducible Research</title>\n",
       "<g id=\"a_node11\"><a xlink:href=\"https://uppsala.instructure.com/courses/51980\" xlink:title=\"Tools for Reproducible Research\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"984.2\" cy=\"-450\" rx=\"77.88\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"984.2\" y=\"-447.8\" font-family=\"Times,serif\" font-size=\"9.00\">Tools for Reproducible Research</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research&#45;&gt;Intro to NGS -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>Tools for Reproducible Research&#45;&gt;Intro to NGS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M963.57,-432.44C940.64,-414.02 903.55,-384.23 878.53,-364.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"870.55,-357.74 881.17,-360.49 874.45,-360.87 878.35,-364 878.35,-364 878.35,-364 874.45,-360.87 875.53,-367.51 870.55,-357.74 870.55,-357.74\"/>\n",
       "</g>\n",
       "<!-- Snakemake BYOC -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>Snakemake BYOC</title>\n",
       "<g id=\"a_node17\"><a xlink:href=\"https://uppsala.instructure.com/courses/52153\" xlink:title=\"Snakemake BYOC\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"984.2\" cy=\"-342\" rx=\"49.15\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"984.2\" y=\"-339.8\" font-family=\"Times,serif\" font-size=\"9.00\">Snakemake BYOC</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research&#45;&gt;Snakemake BYOC -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>Tools for Reproducible Research&#45;&gt;Snakemake BYOC</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M984.2,-431.97C984.2,-415.38 984.2,-389.88 984.2,-370.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"984.2,-360.34 988.7,-370.34 984.2,-365.34 984.2,-370.34 984.2,-370.34 984.2,-370.34 984.2,-365.34 979.7,-370.34 984.2,-360.34 984.2,-360.34\"/>\n",
       "</g>\n",
       "<!-- RNAseq data analysis&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>RNAseq data analysis&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M994.2,-221.35C926.47,-201.63 789.53,-161.75 717.28,-140.71\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"707.53,-137.87 718.39,-136.35 712.33,-139.27 717.13,-140.67 717.13,-140.67 717.13,-140.67 712.33,-139.27 715.87,-144.99 707.53,-137.87 707.53,-137.87\"/>\n",
       "</g>\n",
       "<!-- Spatial transcriptomics -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>Spatial transcriptomics</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"670.2\" cy=\"-18\" rx=\"57.32\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"670.2\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"9.00\" fill=\"gray\">Spatial transcriptomics</text>\n",
       "</g>\n",
       "<!-- single cell RNAseq&#45;&gt;Spatial transcriptomics -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>single cell RNAseq&#45;&gt;Spatial transcriptomics</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M670.2,-107.97C670.2,-91.38 670.2,-65.88 670.2,-46.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"670.2,-36.34 674.7,-46.34 670.2,-41.34 670.2,-46.34 670.2,-46.34 670.2,-46.34 670.2,-41.34 665.7,-46.34 670.2,-36.34 670.2,-36.34\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7fc117d146a0>"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def course_graph2(course_dict_in, output_dir=results_dir, label_type='short',  fontsize='11', fonttype='Arial', node_shape='ellipse'):\n",
    "    \"\"\"Draws a directed graph of course requirements. Same as above, but uses `graphviz` and , and exports the required files.\n",
    "\n",
    "    \n",
    "    Parameters:\n",
    "    ===========\n",
    "    course_dict_in (dict)            a dictionary built from the input yaml file.\n",
    "    output_dir (str)                 output directory\n",
    "    label_type (str)                 node label to show. One of `short` or `long`    \n",
    "    \"\"\"\n",
    "\n",
    "    ## global network setup\n",
    "    g = graphviz.Digraph('G')\n",
    "    g.attr(rankdir='same', compound='true', ranksep = '1', labelloc=\"t\")\n",
    "    fontsize='11'\n",
    "    fonttype='Arial'\n",
    "    node_shape='ellipse'\n",
    "    label_type='short' #node label options: `short`, `long`\n",
    "    g.edge_attr[\"style\"] = \"setlinewidth(.2)\"\n",
    "    g.edge_attr['arrowhead']='vee'\n",
    "    \n",
    "\n",
    "    for course_key in course_in['courses']:\n",
    "        c_att=course_in['courses'][course_key] #course attributes\n",
    "        canvas_link=c_att['canvasLink']\n",
    "        short_title=c_att['shortTitle']\n",
    "        long_title=c_att['longTitle']\n",
    "        prerequisites=c_att['prerequisites']\n",
    "        bioschemas=c_att['bioschemas']        ## used for nothing for now\n",
    "\n",
    "        if(label_type=='short'):\n",
    "            label=short_title\n",
    "            prereq_type='shortTitle'\n",
    "        elif(label_type=='long'):\n",
    "            label=long_title\n",
    "            prereq_type='longTitle'\n",
    "        else:\n",
    "            raise AttributeError('`label` should be one of `short` or `long`.')\n",
    "\n",
    "        # if canvas url is absent show as gray\n",
    "        if(canvas_link==\"\"):\n",
    "            color='gray'\n",
    "            fontsize='9'\n",
    "        else:\n",
    "            color='black'\n",
    "\n",
    "        # create nodes\n",
    "        g.node(label, shape=node_shape, \n",
    "               fontcolor=color, \n",
    "               fontsize=fontsize, \n",
    "               fontstyle=fonttype,\n",
    "               href=canvas_link)\n",
    "        \n",
    "        ## Force \"Intro to Python\" to go to top\n",
    "        if(course_key=='intro_python'):\n",
    "            g.edge('Introduction to Python','Intro to biostatistics and ML', style='invis') \n",
    "\n",
    "        # create edges\n",
    "        if(prerequisites!=[]):\n",
    "            for prerequisite_i in prerequisites:\n",
    "                prereq_name=course_in['courses'][prerequisite_i][prereq_type]\n",
    "                g.edge(prereq_name, label)\n",
    "                \n",
    "    return(g)\n",
    "\n",
    "course_graph2(course_in, results_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4eef362",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def DEPRECATEDcourse_graph(course_dict_in, output_dir=results_dir, label_type='short',  fontsize='11', fonttype='Arial', node_shape='ellipse'):\n",
    "    \"\"\"Draws a directed graph of course requirements, and exports the required files. DOES NOT INCLUDE HYPERLINKS\n",
    "    \n",
    "    Parameters:\n",
    "    ===========\n",
    "    course_dict_in (dict)            a dictionary built from the input yaml file.\n",
    "    output_dir (str)                 output directory\n",
    "    label_type (str)                 node label to show. One of `short` or `long`    \n",
    "    \"\"\"\n",
    "\n",
    "    ## global network setup\n",
    "    g = pgv.AGraph(directed=True, strict=True)\n",
    "\n",
    "    fontsize='11'\n",
    "    fonttype='Arial'\n",
    "    node_shape='ellipse'\n",
    "    label_type='short' #node label options: `short`, `long`\n",
    "    g.edge_attr[\"style\"] = \"setlinewidth(.2)\"\n",
    "    g.edge_attr['arrowhead']='vee'\n",
    "\n",
    "    for course_key in course_in['courses']:\n",
    "        c_att=course_dict_in['courses'][course_key] #course attributes\n",
    "        canvas_link=c_att['canvasLink']\n",
    "        short_title=c_att['shortTitle']\n",
    "        long_title=c_att['longTitle']\n",
    "        prerequisites=c_att['prerequisites']\n",
    "        bioschemas=c_att['bioschemas']        ## used for nothing for now\n",
    "\n",
    "        if(label_type=='short'):\n",
    "            label=short_title\n",
    "            prereq_type='shortTitle'\n",
    "        elif(label_type=='long'):\n",
    "            label=long_title\n",
    "            prereq_type='longTitle'\n",
    "        else:\n",
    "            raise AttributeError('`label` should be one of `short` or `long`.')\n",
    "\n",
    "        # if canvas url is absent show as gray\n",
    "        if(canvas_link==\"\"):\n",
    "            color='gray'\n",
    "            fontsize='9'\n",
    "        else:\n",
    "            color='black'\n",
    "\n",
    "        # create nodes\n",
    "        g.add_node(\n",
    "            label,\n",
    "            label=label,\n",
    "            shape=node_shape, \n",
    "            fontcolor=color, \n",
    "            fontsize=fontsize, \n",
    "            fontstyle=fonttype,\n",
    "            URL=canvas_link\n",
    "        )\n",
    "\n",
    "        # add edges\n",
    "        if(prerequisites!=[]):\n",
    "            for prerequisite_i in prerequisites:\n",
    "                prereq_name=course_dict_in['courses'][prerequisite_i][prereq_type]\n",
    "                g.add_edge(prereq_name, label)\n",
    "\n",
    "    #output\n",
    "    g.layout(prog=\"dot\") \n",
    "    g.draw(output_dir+'course_graph.svg', format='svg')\n",
    "    g.draw(output_dir+'course_graph.map', format='cmapx')\n",
    "    g.draw(output_dir+'course_graph.png')\n",
    "        \n",
    "    return(Image(results_dir+'course_graph.png'))\n",
    "# course_graph(course_in, results_dir)"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:env_training] *",
   "language": "python",
   "name": "conda-env-env_training-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
