{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1712c0d",
   "metadata": {
    "code_folding": [
     20
    ]
   },
   "outputs": [],
   "source": [
    "## Export to html\n",
    "#  jupyter nbconvert course_network.ipynb --no-input --to html\n",
    "# \n",
    "\n",
    "import yaml, re, os, nbconvert, nbformat\n",
    "import pandas as pd\n",
    "import graphviz\n",
    "\n",
    "main_dir=re.sub('code', '', os.getcwd())\n",
    "data_dir=main_dir+'data/'\n",
    "results_dir=main_dir+'results/'\n",
    "\n",
    "with open(data_dir+'course_network.yaml') as file:\n",
    "    course_in=yaml.safe_load(file)\n",
    "    \n",
    "# Using `pygraphviz` for exporting. To export:\n",
    "# 1. jupyter nbconvert course_network.ipynb --no-input --to html\n",
    "# ```\n",
    "# <IMG SRC=\"/path/to/course_graph.svg\" USEMAP=\"#my_graph\" />\n",
    "# ```\n",
    "\n",
    "import pygraphviz as pgv\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "82e22301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Title: Workshop requirements Pages: 1 -->\n",
       "<svg width=\"1427pt\" height=\"476pt\"\n",
       " viewBox=\"0.00 0.00 1427.32 476.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 472)\">\n",
       "<title>Workshop requirements</title>\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-472 1423.32,-472 1423.32,4 -4,4\"/>\n",
       "<!-- Ph.D. Advisory Program -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>Ph.D. Advisory Program</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"75.17\" cy=\"-450\" rx=\"75.33\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"75.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">Ph.D. Advisory Program</text>\n",
       "</g>\n",
       "<!-- Advanced Python -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>Advanced Python</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"323.17\" cy=\"-342\" rx=\"57.25\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"323.17\" y=\"-339.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">Advanced Python</text>\n",
       "</g>\n",
       "<!-- Introduction to Python -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>Introduction to Python</title>\n",
       "<g id=\"a_node3\"><a xlink:href=\"https://uppsala.instructure.com/courses/47059\" xlink:title=\"Introduction to Python\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"238.17\" cy=\"-450\" rx=\"69.51\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"238.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\">Introduction to Python</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Advanced Python -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Advanced Python</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M251.82,-431.97C265.79,-414.56 287.61,-387.34 303.44,-367.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"309.82,-359.64 307.08,-370.26 306.7,-363.54 303.57,-367.44 303.57,-367.44 303.57,-367.44 306.7,-363.54 300.06,-364.63 309.82,-359.64 309.82,-359.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>Intro to biostatistics and ML</title>\n",
       "<g id=\"a_node6\"><a xlink:href=\"https://uppsala.instructure.com/courses/51998\" xlink:title=\"Intro to biostatistics and ML\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"484.17\" cy=\"-342\" rx=\"85.28\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"484.17\" y=\"-339.2\" font-family=\"Times,serif\" font-size=\"11.00\">Intro to biostatistics and ML</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Intro to biostatistics and ML -->\n",
       "<!-- Neural Nets and DL -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>Neural Nets and DL</title>\n",
       "<g id=\"a_node12\"><a xlink:href=\"https://uppsala.instructure.com/courses/47910\" xlink:title=\"Neural Nets and DL\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"391.17\" cy=\"-234\" rx=\"63.28\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"391.17\" y=\"-231.2\" font-family=\"Times,serif\" font-size=\"11.00\">Neural Nets and DL</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Neural Nets and DL -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Neural Nets and DL</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M230.03,-431.83C219.25,-406.34 203.83,-357.89 224.17,-324 247.86,-284.53 295.65,-261.6 334.06,-248.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"343.6,-245.93 335.41,-253.23 338.83,-247.43 334.06,-248.93 334.06,-248.93 334.06,-248.93 338.83,-247.43 332.71,-244.64 343.6,-245.93 343.6,-245.93\"/>\n",
       "</g>\n",
       "<!-- Omics Integration -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>Omics Integration</title>\n",
       "<g id=\"a_node13\"><a xlink:href=\"https://uppsala.instructure.com/courses/52162\" xlink:title=\"Omics Integration\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"576.17\" cy=\"-234\" rx=\"57.96\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"576.17\" y=\"-231.2\" font-family=\"Times,serif\" font-size=\"11.00\">Omics Integration</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M235.42,-431.64C232.27,-405.03 230.54,-354.07 257.17,-324 291.11,-285.67 432.18,-257.6 514.92,-244.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"524.96,-242.45 515.8,-248.48 520.02,-243.24 515.08,-244.04 515.08,-244.04 515.08,-244.04 520.02,-243.24 514.37,-239.6 524.96,-242.45 524.96,-242.45\"/>\n",
       "</g>\n",
       "<!-- single cell RNAseq -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>single cell RNAseq</title>\n",
       "<g id=\"a_node16\"><a xlink:href=\"https://uppsala.instructure.com/courses/52011\" xlink:title=\"single cell RNAseq\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"793.17\" cy=\"-126\" rx=\"61.17\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"793.17\" y=\"-123.2\" font-family=\"Times,serif\" font-size=\"11.00\">single cell RNAseq</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Introduction to Python&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>Introduction to Python&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M228.18,-431.84C215.22,-407.2 195.96,-360.72 211.17,-324 237.14,-261.28 259.13,-247.68 319.17,-216 449.56,-147.2 624.83,-130.94 721.55,-127.49\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"731.81,-127.16 721.96,-131.98 726.81,-127.32 721.81,-127.48 721.81,-127.48 721.81,-127.48 726.81,-127.32 721.67,-122.98 731.81,-127.16 731.81,-127.16\"/>\n",
       "</g>\n",
       "<!-- R Programming -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>R Programming</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"718.17\" cy=\"-450\" rx=\"52.13\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"718.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">R Programming</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Advanced Python -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Advanced Python</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M678.12,-438.25C606.08,-418.92 454.74,-378.31 374.94,-356.89\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"365.1,-354.25 375.93,-352.5 369.93,-355.55 374.76,-356.85 374.76,-356.85 374.76,-356.85 369.93,-355.55 373.6,-361.19 365.1,-354.25 365.1,-354.25\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Intro to biostatistics and ML -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Intro to biostatistics and ML</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M687.58,-435.14C646.78,-416.66 574.31,-383.84 527.8,-362.76\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"518.57,-358.58 529.53,-358.61 523.12,-360.65 527.68,-362.71 527.68,-362.71 527.68,-362.71 523.12,-360.65 525.82,-366.81 518.57,-358.58 518.57,-358.58\"/>\n",
       "</g>\n",
       "<!-- Data visualization with R -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>Data visualization with R</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"665.17\" cy=\"-342\" rx=\"77.65\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"665.17\" y=\"-339.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">Data visualization with R</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Data visualization with R -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Data visualization with R</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M709.65,-431.97C701.15,-414.97 687.98,-388.62 678.17,-369.01\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"673.66,-359.99 682.16,-366.92 675.9,-364.46 678.13,-368.93 678.13,-368.93 678.13,-368.93 675.9,-364.46 674.11,-370.95 673.66,-359.99 673.66,-359.99\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M730.42,-432.44C747.38,-407.34 773.61,-358.87 752.17,-324 726.4,-282.1 674.13,-259.2 633.22,-247.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"623.35,-244.41 634.19,-242.77 628.17,-245.76 632.98,-247.1 632.98,-247.1 632.98,-247.1 628.17,-245.76 631.77,-251.44 623.35,-244.41 623.35,-244.41\"/>\n",
       "</g>\n",
       "<!-- RaukR Summer course -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>RaukR Summer course</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"893.17\" cy=\"-234\" rx=\"71.62\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"893.17\" y=\"-231.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">RaukR Summer course</text>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M740.71,-433.6C763.48,-417.24 798.7,-389.69 823.17,-360 848.55,-329.19 869.62,-287.88 881.81,-261.35\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"886.03,-251.95 886.03,-262.92 883.98,-256.51 881.93,-261.07 881.93,-261.07 881.93,-261.07 883.98,-256.51 877.83,-259.23 886.03,-251.95 886.03,-251.95\"/>\n",
       "</g>\n",
       "<!-- R Programming&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>R Programming&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M731.17,-432.38C743.77,-415.28 762.24,-387.35 771.17,-360 794.48,-288.61 795.6,-198.84 794.42,-154.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"794.08,-144.19 798.91,-154.03 794.24,-149.18 794.41,-154.18 794.41,-154.18 794.41,-154.18 794.24,-149.18 789.91,-154.33 794.08,-144.19 794.08,-144.19\"/>\n",
       "</g>\n",
       "<!-- Bioinformatics for PIs -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>Bioinformatics for PIs</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1350.17\" cy=\"-450\" rx=\"69.31\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1350.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">Bioinformatics for PIs</text>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;Neural Nets and DL -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;Neural Nets and DL</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M469.23,-323.97C453.88,-306.48 429.86,-279.1 412.51,-259.33\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"405.77,-251.64 415.74,-256.19 409.06,-255.4 412.36,-259.16 412.36,-259.16 412.36,-259.16 409.06,-255.4 408.98,-262.12 405.77,-251.64 405.77,-251.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M498.95,-323.97C514.13,-306.48 537.89,-279.1 555.05,-259.33\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"561.73,-251.64 558.57,-262.14 558.45,-255.42 555.17,-259.19 555.17,-259.19 555.17,-259.19 558.45,-255.42 551.77,-256.24 561.73,-251.64 561.73,-251.64\"/>\n",
       "</g>\n",
       "<!-- Intro to biostatistics and ML&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>Intro to biostatistics and ML&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M535.97,-327.58C612.55,-307.73 755.5,-270.68 835.38,-249.98\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"845.28,-247.41 836.72,-254.28 840.43,-248.67 835.59,-249.92 835.59,-249.92 835.59,-249.92 840.43,-248.67 834.47,-245.56 845.28,-247.41 845.28,-247.41\"/>\n",
       "</g>\n",
       "<!-- Epigenomics data analysis -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>Epigenomics data analysis</title>\n",
       "<g id=\"a_node8\"><a xlink:href=\"https://uppsala.instructure.com/courses/51975\" xlink:title=\"Epigenomics data analysis\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1063.17\" cy=\"-234\" rx=\"79.96\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1063.17\" y=\"-231.2\" font-family=\"Times,serif\" font-size=\"11.00\">Epigenomics data analysis</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>Intro to NGS</title>\n",
       "<g id=\"a_node9\"><a xlink:href=\"https://uppsala.instructure.com/courses/47037\" xlink:title=\"Intro to NGS\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1007.17\" cy=\"-342\" rx=\"44.5\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1007.17\" y=\"-339.2\" font-family=\"Times,serif\" font-size=\"11.00\">Intro to NGS</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;Epigenomics data analysis -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;Epigenomics data analysis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M1016.16,-323.97C1025.15,-306.97 1039.06,-280.62 1049.43,-261.01\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1054.19,-251.99 1053.5,-262.93 1051.86,-256.41 1049.52,-260.83 1049.52,-260.83 1049.52,-260.83 1051.86,-256.41 1045.54,-258.73 1054.19,-251.99 1054.19,-251.99\"/>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;Omics Integration -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;Omics Integration</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M970.05,-331.87C893.66,-313.08 718.63,-270.04 629.88,-248.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"620.02,-245.78 630.8,-243.8 624.87,-246.98 629.73,-248.17 629.73,-248.17 629.73,-248.17 624.87,-246.98 628.65,-252.54 620.02,-245.78 620.02,-245.78\"/>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;RaukR Summer course -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;RaukR Summer course</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M989.85,-324.9C970.88,-307.26 940.26,-278.79 918.57,-258.62\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"911.02,-251.6 921.41,-255.11 914.68,-255.01 918.34,-258.41 918.34,-258.41 918.34,-258.41 914.68,-255.01 915.28,-261.71 911.02,-251.6 911.02,-251.6\"/>\n",
       "</g>\n",
       "<!-- RNAseq data analysis -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>RNAseq data analysis</title>\n",
       "<g id=\"a_node15\"><a xlink:href=\"https://uppsala.instructure.com/courses/52164\" xlink:title=\"RNAseq data analysis\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1230.17\" cy=\"-234\" rx=\"68.61\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1230.17\" y=\"-231.2\" font-family=\"Times,serif\" font-size=\"11.00\">RNAseq data analysis</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to NGS&#45;&gt;RNAseq data analysis -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>Intro to NGS&#45;&gt;RNAseq data analysis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M1034.96,-327.79C1073.76,-309.34 1144.45,-275.75 1189.27,-254.44\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1198.42,-250.09 1191.32,-258.45 1193.9,-252.24 1189.39,-254.38 1189.39,-254.38 1189.39,-254.38 1193.9,-252.24 1187.46,-250.32 1198.42,-250.09 1198.42,-250.09\"/>\n",
       "</g>\n",
       "<!-- Intro to Data Management -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>Intro to Data Management</title>\n",
       "<g id=\"a_node10\"><a xlink:href=\"https://uppsala.instructure.com/courses/52280\" xlink:title=\"Intro to Data Management\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"971.17\" cy=\"-450\" rx=\"79.96\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"971.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\">Intro to Data Management</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Intro to Data Management&#45;&gt;Intro to NGS -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>Intro to Data Management&#45;&gt;Intro to NGS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M976.95,-431.97C982.67,-415.12 991.51,-389.1 998.15,-369.54\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1001.4,-359.99 1002.44,-370.9 999.79,-364.72 998.18,-369.46 998.18,-369.46 998.18,-369.46 999.79,-364.72 993.92,-368.01 1001.4,-359.99 1001.4,-359.99\"/>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>Tools for Reproducible Research</title>\n",
       "<g id=\"a_node11\"><a xlink:href=\"https://uppsala.instructure.com/courses/51980\" xlink:title=\"Tools for Reproducible Research\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1166.17\" cy=\"-450\" rx=\"96.63\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1166.17\" y=\"-447.2\" font-family=\"Times,serif\" font-size=\"11.00\">Tools for Reproducible Research</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research&#45;&gt;Intro to NGS -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>Tools for Reproducible Research&#45;&gt;Intro to NGS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M1141.32,-432.44C1113.34,-413.78 1067.85,-383.46 1037.72,-363.37\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1029.28,-357.74 1040.09,-359.54 1033.44,-360.51 1037.6,-363.29 1037.6,-363.29 1037.6,-363.29 1033.44,-360.51 1035.1,-367.03 1029.28,-357.74 1029.28,-357.74\"/>\n",
       "</g>\n",
       "<!-- Snakemake BYOC -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>Snakemake BYOC</title>\n",
       "<g id=\"a_node17\"><a xlink:href=\"https://uppsala.instructure.com/courses/52153\" xlink:title=\"Snakemake BYOC\" target=\"_blank\">\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1166.17\" cy=\"-342\" rx=\"60.97\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1166.17\" y=\"-339.2\" font-family=\"Times,serif\" font-size=\"11.00\">Snakemake BYOC</text>\n",
       "</a>\n",
       "</g>\n",
       "</g>\n",
       "<!-- Tools for Reproducible Research&#45;&gt;Snakemake BYOC -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>Tools for Reproducible Research&#45;&gt;Snakemake BYOC</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M1166.17,-431.97C1166.17,-415.38 1166.17,-389.88 1166.17,-370.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"1166.17,-360.34 1170.67,-370.34 1166.17,-365.34 1166.17,-370.34 1166.17,-370.34 1166.17,-370.34 1166.17,-365.34 1161.67,-370.34 1166.17,-360.34 1166.17,-360.34\"/>\n",
       "</g>\n",
       "<!-- RNAseq data analysis&#45;&gt;single cell RNAseq -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>RNAseq data analysis&#45;&gt;single cell RNAseq</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M1181.73,-221.25C1099.97,-201.42 935.06,-161.42 848.73,-140.48\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"838.85,-138.08 849.63,-136.07 843.71,-139.26 848.57,-140.44 848.57,-140.44 848.57,-140.44 843.71,-139.26 847.51,-144.81 838.85,-138.08 838.85,-138.08\"/>\n",
       "</g>\n",
       "<!-- Spatial transcriptomics -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>Spatial transcriptomics</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"793.17\" cy=\"-18\" rx=\"70.92\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"793.17\" y=\"-15.2\" font-family=\"Times,serif\" font-size=\"11.00\" fill=\"gray\">Spatial transcriptomics</text>\n",
       "</g>\n",
       "<!-- single cell RNAseq&#45;&gt;Spatial transcriptomics -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>single cell RNAseq&#45;&gt;Spatial transcriptomics</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"0.2\" d=\"M793.17,-107.97C793.17,-91.38 793.17,-65.88 793.17,-46.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"0.2\" points=\"793.17,-36.34 797.67,-46.34 793.17,-41.34 793.17,-46.34 793.17,-46.34 793.17,-46.34 793.17,-41.34 788.67,-46.34 793.17,-36.34 793.17,-36.34\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7fe21f6220d0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def course_graph2(course_dict_in, output_dir=results_dir, label_type='short',  font_size='11', fonttype='Arial', node_shape='ellipse'):\n",
    "    \"\"\"Draws a directed graph of course requirements. Same as above, but uses `graphviz` and , and exports the required files.\n",
    "\n",
    "    \n",
    "    Parameters:\n",
    "    ===========\n",
    "    course_dict_in (dict)            a dictionary built from the input yaml file.\n",
    "    output_dir (str)                 output directory\n",
    "    label_type (str)                 node label to show. One of `short` or `long`    \n",
    "    \"\"\"\n",
    "\n",
    "    ## global network setup\n",
    "    g = graphviz.Digraph('Workshop requirements')\n",
    "    g.attr(rankdir='same', compound='true', ranksep = '1', labelloc=\"t\")\n",
    "    fonttype='Arial'\n",
    "    node_shape='ellipse'\n",
    "    label_type='short' #node label options: `short`, `long`\n",
    "    g.node_attr[\"target\"]=\"_blank\"\n",
    "    g.edge_attr[\"style\"] = \"setlinewidth(.2)\"\n",
    "    g.edge_attr['arrowhead']='vee'\n",
    "    \n",
    "\n",
    "    for course_key in course_in['courses']:\n",
    "        c_att=course_in['courses'][course_key] #course attributes\n",
    "        canvas_link=c_att['canvasLink']\n",
    "        short_title=c_att['shortTitle']\n",
    "        long_title=c_att['longTitle']\n",
    "        prerequisites=c_att['prerequisites']\n",
    "        bioschemas=c_att['bioschemas']        ## used for nothing for now\n",
    "\n",
    "        if(label_type=='short'):\n",
    "            label=short_title\n",
    "            prereq_type='shortTitle'\n",
    "        elif(label_type=='long'):\n",
    "            label=long_title\n",
    "            prereq_type='longTitle'\n",
    "        else:\n",
    "            raise AttributeError('`label` should be one of `short` or `long`.')\n",
    "\n",
    "        # if canvas url is absent show as gray\n",
    "        if(canvas_link==\"\"):\n",
    "            color='gray'\n",
    "#             fsize='9'\n",
    "        else:\n",
    "            color='black'\n",
    "\n",
    "        # create nodes\n",
    "        g.node(label, shape=node_shape, \n",
    "               fontcolor=color, \n",
    "               fontsize=font_size, \n",
    "               fontstyle=fonttype,\n",
    "               href=canvas_link)\n",
    "        \n",
    "        ## Force \"Intro to Python\" to go to top\n",
    "        if(course_key=='intro_python'):\n",
    "            g.edge('Introduction to Python','Intro to biostatistics and ML', style='invis') \n",
    "\n",
    "        # create edges\n",
    "        if(prerequisites!=[]):\n",
    "            for prerequisite_i in prerequisites:\n",
    "                prereq_name=course_in['courses'][prerequisite_i][prereq_type]\n",
    "                g.edge(prereq_name, label)\n",
    "                \n",
    "    return(g)\n",
    "\n",
    "course_graph2(course_in, results_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4eef362",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def DEPRECATEDcourse_graph(course_dict_in, output_dir=results_dir, label_type='short',  fontsize='11', fonttype='Arial', node_shape='ellipse'):\n",
    "    \"\"\"Draws a directed graph of course requirements, and exports the required files. DOES NOT INCLUDE HYPERLINKS\n",
    "    \n",
    "    Parameters:\n",
    "    ===========\n",
    "    course_dict_in (dict)            a dictionary built from the input yaml file.\n",
    "    output_dir (str)                 output directory\n",
    "    label_type (str)                 node label to show. One of `short` or `long`    \n",
    "    \"\"\"\n",
    "\n",
    "    ## global network setup\n",
    "    g = pgv.AGraph(directed=True, strict=True)\n",
    "\n",
    "    fontsize='11'\n",
    "    fonttype='Arial'\n",
    "    node_shape='ellipse'\n",
    "    label_type='short' #node label options: `short`, `long`\n",
    "    g.edge_attr[\"style\"] = \"setlinewidth(.2)\"\n",
    "    g.edge_attr['arrowhead']='vee'\n",
    "\n",
    "    for course_key in course_in['courses']:\n",
    "        c_att=course_dict_in['courses'][course_key] #course attributes\n",
    "        canvas_link=c_att['canvasLink']\n",
    "        short_title=c_att['shortTitle']\n",
    "        long_title=c_att['longTitle']\n",
    "        prerequisites=c_att['prerequisites']\n",
    "        bioschemas=c_att['bioschemas']        ## used for nothing for now\n",
    "\n",
    "        if(label_type=='short'):\n",
    "            label=short_title\n",
    "            prereq_type='shortTitle'\n",
    "        elif(label_type=='long'):\n",
    "            label=long_title\n",
    "            prereq_type='longTitle'\n",
    "        else:\n",
    "            raise AttributeError('`label` should be one of `short` or `long`.')\n",
    "\n",
    "        # if canvas url is absent show as gray\n",
    "        if(canvas_link==\"\"):\n",
    "            color='gray'\n",
    "            fontsize='9'\n",
    "        else:\n",
    "            color='black'\n",
    "\n",
    "        # create nodes\n",
    "        g.add_node(\n",
    "            label,\n",
    "            label=label,\n",
    "            shape=node_shape, \n",
    "            fontcolor=color, \n",
    "            fontsize=fontsize, \n",
    "            fontstyle=fonttype,\n",
    "            URL=canvas_link\n",
    "        )\n",
    "\n",
    "        # add edges\n",
    "        if(prerequisites!=[]):\n",
    "            for prerequisite_i in prerequisites:\n",
    "                prereq_name=course_dict_in['courses'][prerequisite_i][prereq_type]\n",
    "                g.add_edge(prereq_name, label)\n",
    "\n",
    "    #output\n",
    "    g.layout(prog=\"dot\") \n",
    "    g.draw(output_dir+'course_graph.svg', format='svg')\n",
    "    g.draw(output_dir+'course_graph.map', format='cmapx')\n",
    "    g.draw(output_dir+'course_graph.png')\n",
    "        \n",
    "    return(Image(results_dir+'course_graph.png'))\n",
    "# course_graph(course_in, results_dir)"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:env_training] *",
   "language": "python",
   "name": "conda-env-env_training-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
